src/smt/process_assertions.cpp

```diff
@@ -340,10 +340,8 @@ bool ProcessAssertions::apply(Assertions& as)
   d_passes["theory-rewrite-eq"]->apply(&assertions);
   // apply theory preprocess, which includes ITE removal
   d_passes["theory-preprocess"]->apply(&assertions);
-  // This is needed because when solving incrementally, removeITEs may
-  // introduce skolems that were solved for earlier and thus appear in the
-  // substitution map.
-  d_passes["apply-substs"]->apply(&assertions);
+  // notice that we do not apply substitutions as a last step here, since
+  // the range of substitutions is not theory-preprocessed.
```

Comment:
Are we applying the substitutions somewhere else now?

Reply:
Substitutions have always been applied locally within non-clausal simplification (https://github.com/CVC4/CVC4/blob/master/src/preprocessing/passes/non_clausal_simp.cpp#L302), which appears to be the only place that creates substitutions.

This call to "apply-subst" (which takes the global map of substitutions) appears to me to have been paranoia and designed for cases where skolems introduced by theory-preprocessing could be solved for.  However, it is highly counter-intuitive that skolems would be solved for. I believe theory-preprocessing should go last, implying that skolems are "private" to theory engine / prop engine.  If we do solve for skolems in preprocessing, this would be very rare.

Regardless, I admit this PR is a pretty tricky change that we should test.